hw02-generics ready to check

limentur opened the pull request:
hw02-generics ready to check

limentur committed:
hw02-generics ready to check

spvgl commented on L03-generics/src/main/java/homework/Customer.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    public int hashCode() {
+        //        result = 31 * result + (name != null ? name.hashCode() : 0);
+//        result = 31 * result + Long.hashCode(scores);
```
Закомментированный код лучше не комитить.

spvgl commented on L03-generics/src/main/java/homework/Customer.java:
```diff
@@ -0,0 +1,63 @@
+
+        Customer customer = (Customer) o;
+
+        if (id != customer.id) return false;
+        if (scores != customer.scores) return false;
+        return Objects.equals(name, customer.name);
```
Подход к сравнению в `equals` и к вычислению хэша в `hashCode` должны быть одинаковые (то, от каких полей зависит результат).

spvgl commented on L03-generics/src/main/java/homework/CustomerReverseOrder.java:
```diff
@@ -0,0 +1,24 @@
+
+public class CustomerReverseOrder {
+
+    private List<Customer> reversedList = new LinkedList<>();
+    private boolean isReversed = false;
+
+    public void add(Customer customer) {
+        reversedList.add(customer);
+    }
+
+    public Customer take() {
+        Customer lastCustomer;
+        if (!isReversed){
+            reversedList = reversedList.reversed();
+            isReversed = true;
+        }
+        lastCustomer =  reversedList.getFirst();
+        reversedList.remove(reversedList.getFirst());
+        return lastCustomer;
+    }
```
Интересное решение. Посмотрите на интерфейс `Deque` -- с ним решение будет в одну строку.

spvgl commented on L03-generics/src/main/java/homework/CustomerService.java:
```diff
@@ -0,0 +1,61 @@
+
+    private final Map<Customer, String> customerMap = new HashMap<>();
+
+    public Map.Entry<Customer, String> getSmallest() {
+        Map.Entry<Customer, String> min = null;
+        for (Map.Entry<Customer, String> entryMap : customerMap.entrySet()) {
+            if (min == null || min.getKey().getScores() > entryMap.getKey().getScores()) {
+                min = new AbstractMap.SimpleEntry<>(entryMap.getKey(), entryMap.getValue());
+            }
+        }
+
+        assert min != null;
+        return copyOfEntry(min);
+    }
+
+    public Map.Entry<Customer, String> getNext(Customer customer) {
+        Map.Entry<Customer, String> middle = null;
+        for (Map.Entry<Customer, String> entryMap : customerMap.entrySet()) {
+            if (middle == null || (middle.getKey().getScores() < entryMap.getKey().getScores()) && middle.getKey().getScores() < customer.getScores()) {
+                middle = entryMap;
+            }
+        }
+
+        assert middle != null;
+        if (customer.getScores() > middle.getKey().getScores()){
+            return null;
+        }
+
+        return middle;
+    }
```
Посмотрите на интерфейс `NavigableMap`. Перебора и циклов здесь не должно быть.

limentur committed:
hw02-generics ready to check after fixes

spvgl commented on L03-generics/src/main/java/homework/Customer.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        Customer c = (Customer) o;
+        if (c.getId() == id) return true;
+        return getClass() == o.getClass();
+    }
```
Очень странно выглядит сравнение по `getClass()` после сравнения по `id`.
Подумайте, что будет если передать в ваш `equals` не `Customer`, а другой класс?

Должно быть примерно так:
```java
  if (this == o) return true;
  if (getClass() != o.getClass()) return false;
  Customer c = (Customer) o;
  return c.getId() == id;
```
Только надо еще проверку добавить. Какую?

limentur committed:
hw02-generics ready to check after fixes
added check on null in Customer.class
without it code can invoke Object.getClass of null and drop an error
